Modules/_testcapimodule.c

```diff
@@ -5169,6 +5169,226 @@ test_tstate_capi(PyObject *self, PyObject *Py_UNUSED(args))
+        return -1;
+    } else if (!PyErr_Occurred()) {
+        raiseTestError("test_watch_dict", "PyDict_Watch() returned error code without exception set");
+        return -1;
+    } else {
+        PyErr_Clear();
+    }
+    return 0;
+}
+
+static PyObject *
+test_watch_dict(PyObject *self, PyObject *Py_UNUSED(args))
+{
+    PyObject *watched = PyDict_New();
+    PyObject *unwatched = PyDict_New();
+    PyObject *one = PyLong_FromLong(1);
+    PyObject *two = PyLong_FromLong(2);
+    PyObject *key1 = PyUnicode_FromString("key1");
+    PyObject *key2 = PyUnicode_FromString("key2");
+
+    g_dict_watch_events = PyList_New(0);
+
+    int wid = PyDict_AddWatcher(dict_watch_callback);
+    if (try_watch(wid, watched)) {
+        return NULL;
+    }
+
+    PyDict_SetItem(unwatched, key1, two);
+    PyDict_Merge(watched, unwatched, 1);
+
+    if (dict_watch_assert(1, "clone")) {
+        return NULL;
+    }
+
+    PyDict_SetItem(watched, key1, one);
+    PyDict_SetItem(unwatched, key1, one);
+
+    if (dict_watch_assert(2, "mod:key1:1")) {
+        return NULL;
+    }
+
+    PyDict_SetItemString(watched, "key1", two);
+    PyDict_SetItemString(unwatched, "key1", two);
+
+    if (dict_watch_assert(3, "mod:key1:2")) {
+        return NULL;
+    }
+
+    PyDict_SetItem(watched, key2, one);
+    PyDict_SetItem(unwatched, key2, one);
+
+    if (dict_watch_assert(4, "new:key2:1")) {
+        return NULL;
+    }
+
+    _PyDict_Pop(watched, key2, Py_None);
+    _PyDict_Pop(unwatched, key2, Py_None);
+
+    if (dict_watch_assert(5, "del:key2")) {
+        return NULL;
+    }
+
+    PyDict_DelItemString(watched, "key1");
+    PyDict_DelItemString(unwatched, "key1");
+
+    if (dict_watch_assert(6, "del:key1")) {
+        return NULL;
+    }
+
+    PyDict_SetDefault(watched, key1, one);
+    PyDict_SetDefault(unwatched, key1, one);
+
+    if (dict_watch_assert(7, "new:key1:1")) {
+        return NULL;
+    }
+
+    int wid2 = PyDict_AddWatcher(dict_watch_callback_2);
+    if (try_watch(wid2, unwatched)) {
+        return NULL;
+    }
+
+    PyDict_Clear(watched);
+
+    if (dict_watch_assert(8, "clear")) {
+        return NULL;
+    }
+
+    PyDict_Clear(unwatched);
+
+    if (dict_watch_assert(9, "second")) {
+        return NULL;
+    }
+
+    PyObject *copy = PyDict_Copy(watched);
+    // copied dict is not watched, so this does not add an event
+    Py_CLEAR(copy);
+
+    Py_CLEAR(watched);
+
+    if (dict_watch_assert(10, "dealloc")) {
+        return NULL;
+    }
+
+    // it is an error to try to watch a non-dict
+    if (dict_watch_assert_error(wid, one, "PyDict_Watch() succeeded on non-dict")) {
+        return NULL;
+    }
+
+    // It is an error to pass an out-of-range watcher ID
+    if (dict_watch_assert_error(-1, unwatched, "PyDict_Watch() succeeded on negative watcher ID")) { 
+        return NULL;
+    }
+    if (dict_watch_assert_error(8, unwatched, "PyDict_Watch() succeeded on too-large watcher ID")) {
```

Comment:
Change 8 to `DICT_MAX_WATCHERS`?

Reply:
When I implemented the other comment about moving the bulk of the tests into Python, I lost access to the C-defined constant, so I couldn't do this. But I added a comment clarifying the origin of the magic number.